File: go/tree/inorderSuccessor.go
package tree

func inorderSuccessor(root *TreeNode, p *TreeNode) *TreeNode {
	curr := root
	var result *TreeNode = nil
	for curr != nil {
		if curr.Val > p.Val {
			result = curr
			curr = curr.Left
		} else {
			curr = curr.Right
		}
	}
	return result
}

File: java/src/main/java/io/dure/tree/InorderSuccessor.java
package io.dure.tree;

public class InorderSuccessor {
    public TreeNode inorderSuccessor(TreeNode root, TreeNode p) {
        TreeNode curr = root;
        TreeNode result = null;
        while (curr != null) {
            if (curr.val > p.val) {
                result = curr;
                curr = curr.left;
            } else {
                curr = curr.right;
            }
        }
        return result;
    }
}
